gotsh/gotsh.1
.\"
.\" Copyright (c) 2022 Stefan Sperling
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt GOTSH 1
.Os
.Sh NAME
.Nm gotsh
.Nd Game of Trees Shell
.Sh SYNOPSIS
.Nm Fl c Sq Cm git-receive-pack Ar repository-path
.Nm Fl c Sq Cm git-upload-pack Ar repository-path
.Nm Fl c Sq Cm weblogin Oo Ar hostname Oc
.Sh DESCRIPTION
.Nm
is the network-facing interface to
.Xr gotd 8 .
It implements the server-side part of the Git network protocol used by
.Xr git 1
and
.Xr got 1 .
.Pp
.Nm
is not an interactive shell.
.Nm
is intended to be configured as the login shell of Git repository
user accounts on servers running
.Xr gotd 8 .
If users require a different login shell,
.Nm
can be installed in the command search path under the names
.Cm git-receive-pack
and
.Cm git-upload-pack ,
or
.Xr gitwrapper 1
can be used to select the appropriate command to run automatically.
.Pp
The users can then interact with
.Xr gotd 8
over the network.
When users invoke commands such as
.Cm got send
and
.Cm got fetch
on client machines,
.Xr got 1
will connect to the server with
.Xr ssh 1 .
.Nm
will facilitate communication between
.Xr gotd 8
running on the server machine and the
.Xr got 1
or
.Xr git 1
program running on the client machine.
.Pp
Users running
.Nm
should not have access to Git repositories by means other than
accessing the unix socket of
.Xr gotd 8
via
.Nm .
.Pp
The
.Cm weblogin
command provides user authentication for
.Xr gotwebd 8 .
.Nm
will connect to
.Xr gotwebd 8
and obtain a login URL which allows browsing private repositories the user
has been granted read access to in
.Xr gotwebd.conf 5 .
If multiple servers are declared in
.Xr gotwebd.conf 5
the
.Ar hostname
parameter is required and indicates the desired virtual host to use in the URL.
If no
.Ar hostname
is specified and only one server is declared in
.Xr gotwebd.conf 5
then the name of this server will be used in the URL.
.Pp
It is recommended to restrict
.Xr ssh 1
features available to users of
.Nm .
See the
.Sx EXAMPLES
section for details.
.Sh ENVIRONMENT
.Bl -tag -width GOTD_UNIX_SOCKET
.It Ev GOTD_UNIX_SOCKET
Set the path to the unix socket which
.Xr gotd 8
is listening on.
If not specified, the default path
.Pa /var/run/gotd.sock
will be used.
.El
.Sh EXAMPLES
.Xr sshd_config 5
directives such as the following are recommended to protect the server
machine and any systems reachable from it, especially if anonymous users
are allowed to connect:
.Bd -literal -offset indent
Match User developer
    DisableForwarding yes
    PermitTTY no
.Ed
.Pp
It can be convenient to add all relevant users to a common group, such as
.Dq developers ,
and then use this group as the Match criteria:
.Bd -literal -offset indent
Match Group developers
    DisableForwarding yes
    PermitTTY no
.Ed
.Pp
Anonymous users can be given public read-only access by using a
.Xr gotd.conf 5
access rule such as the following:
.Bd -literal -offset indent
repository "public" {
	path "/var/git/public.git"
	permit ro anonymous
}
.Ed
.Pp
The anonymous user account should have a publicly known password, or can be
set up with an empty password in which case the user's
.Xr vipw 8
entry would look similar to this example:
.Bd -literal
anonymous::1002:1002::0:0:Anonymous:/home/anonymous:/usr/local/bin/gotsh
.Ed
.Pp
Use of an empty password must be explicitly allowed in
.Xr sshd_config 5 :
.Bd -literal -offset indent
Match User anonymous
    PasswordAuthentication yes
    PermitEmptyPasswords yes
    DisableForwarding yes
    PermitTTY no
.Ed
.Pp
And to tell
.Xr security 8
to ignore the empty password, set PASSWDSKIP to the following in
.Pa /etc/daily.local :
.Bd -literal -offset indent
PASSWDSKIP="anonymous:/usr/local/bin/gotsh"
.Ed
.Pp
Obtain a
.Xr gotwebd 8
login URL for got.example.com:
.Bd -literal -offset indent
$ ssh got.example.com weblogin
.Ed
.Pp
If the web server at got.example.com serves virtual hosts then two
hostnames must be provided.
One for
.Xr ssh 1
to connect to, and another to identify the virtual host served by
.Xr gotwebd 8 :
.Bd -literal -offset indent
$ ssh got.example.com weblogin got.example.com
.Ed
.Pp
In practice both hostnames will often be the same, but this is not guaranteed.
There is no reliable way determine the desired virtual host automatically.
An
.Xr ssh_config 5
entry like the following can save some typing:
.Bd -literal -offset indent
Host weblogin
	Hostname got.example.com
	RemoteCommand weblogin %h
.Ed
.Pp
The following command is now equivalent to the above:
.Bd -literal -offset indent
$ ssh weblogin
.Ed
.Sh SEE ALSO
.Xr gitwrapper 1 ,
.Xr got 1 ,
.Xr ssh 1 ,
.Xr gotd.conf 5 ,
.Xr gotwebd.conf 5 ,
.Xr sshd_config 5 ,
.Xr gotd 8 ,
.Xr gotwebd 8 ,
.Xr security 8
.Sh AUTHORS
.An Stefan Sperling Aq Mt stsp@openbsd.org


CHANGES
* got 0.120; 2025-10-03
- disable gotwebd authentication if it is not enabled in /etc/gotwebd.conf
- ensure that GOTWEBD_LOGIN_TIMEOUT is used consistently at build time
- prevent date-specific gotsysd regress failures due to asctime_r whitespace
- make gotwebd refuse to start up if the _gotwebd user is root
- make gotwebd warn if the webserver's user is set to root in /etc/gotwebd.conf
- add /etc/gotwebd.conf parameters for hiding repositories
- reject bad hostnames provided to the gotsh weblogin command
- allow gotwebd to optionally display a login hint when authentication fails

* got 0.119; 2025-09-29
- clean up gotwebd requests before freeing sockets; avoids deref of NULL sock
- fix wrong process names in gotwebd error messages
- use an absolute URL in redirects generated by gotwebd
- fix an issue where histedit would not merge added files correctly
- tweak gotwebd deferred accept handling to work more like httpd's
- add a 'weblogin' command to gotsh for use with gotwebd user authentication
- add user authentication support to gotwebd
- make gotwebd IPC sockets use non-blocking I/O as intended to fix hang
- prevent gotd looping on unhandled errors raised by accept()
- fix gotsysd.conf "listen on" statement failing on non-existent sockets
- pass gotwebd.conf's www user setting to every gotwebd child process

* got 0.118; 2025-09-09
- security fix for -portable: gotwebd can be tricked into reading repositories
  outside its repos_path; bug introduced in got-0.111; OpenBSD is not affected
- make 'tog diff' show the repository name in names of patches written to /tmp
- plug memory leaks which were making gotwebd regress tests fail
- fix parallel processing of requests in gotwebd, improving responsiveness
- set gotwebd pledges according to address families of listening sockets
- run gotwebd fcgi parameter parsing in a dedicated process under pledge "stdio"
- make gotd commit notifications only show history which is unique to the branch
- enable sftp/scp support in the sshd_config file generated by gotsysd
- make gotsysd-managed repositories readable for the _gotd group

* got 0.117; 2025-08-17
- regress: replace "sed -i" with ed(1) for portable in-place editing
- ensure that error messages from gotsysd libexec helpers get logged
- fix gotsysd using wrong auth and hmac labels in the generated gotd.conf
- preserve bad symlinks across merges during rebase and histedit
- improve binary files detection: detect any control characters, not just NUL
- gotwebd: fix race condition resulting in trucated html with trailing garbage
- make commit coloring faster and more accurate, producing smaller pack files
- improve selection of pack files for pinning in the open pack file cache
- regress: don't load global/home git configuration files while running tests
- make 'got clone' set a got.conf default branch for fetching only, not sending

* got 0.116; 2025-07-25
- make our pack-refs header format align with the expectations of git 2.50.0
- fix bogus "bad offset in pack file" errors wrongly raised by gotd
- fix gotd branch protection rejecting commits that already exist on server
- pick a default branch to clone when the server does not advertise HEAD symref
- do not clobber changes staged via stage -p during "got revert"
- enforce additional restrictions on reference names specified in gotsys.conf
- change gotwebd favicons to show the smiley fish only
- fix gotd reload when /etc/gotd-secrets.conf is used
- fix bogus "raw object has unexpected size" errors during deltification
- fix bug in delta block stretch size calculation resulting in invalid deltas
- fix gotsysd behaviour when the anonymous user is removed from gotsys.conf
- add support for email and http/json notifications to gotsysd and gotsys.conf

* got 0.115; 2025-06-27
- make errors reported by gotsys-apply-conf actually visible
- stop trying to start gotd from gotsys-apply-conf if gotd is not running
- fix infinite loop in got_pack_repaint_parent_commits() and got-read-pack
- fix creation of gotd.conf deny rules in gotsys-write-conf
- add support for global repository access rules to gotsysd.conf
- fix segfault due to double-free in got-read-gotconfig

* got 0.114; 2025-06-23
- preserve author timestamps when rebasing commits
- stop running ssh with -q by default; -q hides host key fingerprint errors
- fix gotsys-read-conf crash when ssh key comments are missing in gotsys.conf
- relax repository path permission checks in gotsys-repo-create
- add gotsys apply -w option which waits until sysconf has been run
- fix gotsysd getting stuck due to missing final messages from libexec helpers
- plug a file descriptor leak in the gotsysd libexec process

* got 0.113; 2025-06-05
- tweak 'got status' and 'got add' ignores handling for better git compatibility
- improve redundant pack detection during 'gotadmin cleanup'
- gotwebd: do not forget to initialize *ngroups argument for getgrouplist()
- fix default access for root and _gotd when gotsysd runs without config file
- fix bogus "unexpected privsep message" from gotsh during 'got send'
- fix a race in gotd notification processing causing notify process to exit

* got 0.112; 2025-05-14
- remove /tmp/got-importmsg temp files when import commit message is left empty
- rely on secondary _gotwebd groups if repos_path is not owned by _gotwebd group
- fix unrelated errors being reported if a histedit operation is aborted
- implement support for protected references in gotsys.conf and gotsysd
- plug memory leaks in some libexec helpers and in the gitconfig parser
- stop needlessly opening the repository whenever a work tree is opened

* got 0.111; 2025-04-22
- introduce gotsysd: configure gotd servers by committing to gotsys.git repo
- make gotd run 'gotsys check' on gotsys.conf commits before accepting them
- make gotd run 'gotsys apply' when the gotsys.git repo receives changes
- add a missing malloc failure check to gotd's repo_write process

gotsh/gotsh.c
/*
 * Copyright (c) 2022 Stefan Sperling <stsp@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/queue.h>
#include <sys/tree.h>
#include <sys/socket.h>
#include <sys/un.h>

#include <ctype.h>
#include <err.h>
#include <event.h>
#include <imsg.h>
#include <limits.h>
#include <sha1.h>
#include <sha2.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "got_error.h"
#include "got_object.h"
#include "got_serve.h"
#include "got_path.h"
#include "got_reference.h"

#include "got_lib_dial.h"
#include "got_lib_poll.h"

#include "gotd.h"
#include "gotwebd.h"

static int chattygot;

__dead static void
usage(void)
{
	fprintf(stderr, "usage: %s -c '%s|%s repository-path'\n",
	    getprogname(), GOT_DIAL_CMD_SEND, GOT_DIAL_CMD_FETCH);
	fprintf(stderr, "       %s -c 'weblogin [hostname]'\n", getprogname());
	exit(1);
}

static const struct got_error *
apply_unveil(const char *unix_socket_path)
{
#ifdef PROFILE
	if (unveil("gmon.out", "rwc") != 0)
		return got_error_from_errno2("unveil", "gmon.out");
#endif
	if (unveil(unix_socket_path, "w") != 0)
		return got_error_from_errno2("unveil", unix_socket_path);

	if (unveil(NULL, NULL) != 0)
		return got_error_from_errno("unveil");

	return NULL;
}

/* Read session URL from gotwebd's auth socket and send it to the client. */
static const struct got_error *
weblogin(FILE *out, int sock, const char *hostname)
{
	const struct got_error *err = NULL;
	FILE *fp;
	int ret;
	char *line = NULL;
	size_t linesize;
	ssize_t linelen;

	fp = fdopen(sock, "w+");
	if (fp == NULL)
		return got_error_from_errno("fdopen");

	ret = fprintf(fp, "login%s%s\n", hostname != NULL ? " " : "",
	    hostname != NULL ? hostname : "");
	if (ret < 0) {
		err = got_error_from_errno("fprintf");
		goto done;
	}

	/*
	 * gotwebd will return "ok URL", likewise terminated by \n,
	 * or might return an arbitrary error message + \n.
	 * We don't know how long this line will be, so keep reading
	 * in chunks until we have read all of it.
	 * For forward compatibilty, ignore any trailing lines received.
	 */
	linelen = getline(&line, &linesize, fp);
	if (linelen == -1) {
		err = got_error(GOT_ERR_EOF);
		if (ferror(fp))
			err = got_error_from_errno("getline");
		goto done;
	}

	if (strncmp(line, "ok ", 3) == 0) {
		fprintf(out, "Login successful.  Please visit the following "
		    "URL within the next %d minutes: %s\n",
		    GOTWEBD_LOGIN_TIMEOUT / 60, line + 3);
		goto done;
	}

	if (strncmp(line, "err ", 4) == 0) {
		err = got_error_fmt(GOT_ERR_LOGIN_FAILED, "%s", line + 4);
		goto done;
	}

	err = got_error(GOT_ERR_UNKNOWN_COMMAND);
done:
	if (line != NULL)
		free(line);
	if (fp != NULL && fclose(fp) == EOF && err == NULL)
		err = got_error_from_errno("fclose");
	return err;
}


static const struct got_error *
parse_weblogin_command(char **hostname, char *cmd)
{
	size_t len, cmdlen;

	*hostname = NULL;

	len = strlen(cmd);

	while (len > 0 && isspace(cmd[len - 1]))
		cmd[--len] = '\0';

	if (len == 0)
		return got_error(GOT_ERR_BAD_PACKET);

	if (len >= strlen(GOTWEBD_LOGIN_CMD) &&
	    strncmp(cmd, GOTWEBD_LOGIN_CMD, strlen(GOTWEBD_LOGIN_CMD)) == 0)
		cmdlen = strlen(GOTWEBD_LOGIN_CMD);
	else
		return got_error(GOT_ERR_BAD_PACKET);

	/* The hostname parameter is optional. */
	if (len == cmdlen)
		return NULL;

	if (len <= cmdlen + 1 || cmd[cmdlen] != ' ')
		return got_error(GOT_ERR_BAD_PACKET);

	if (memchr(&cmd[cmdlen + 1], '\0', len - cmdlen) == NULL)
		return got_error(GOT_ERR_BAD_PACKET);

	/* Forbid linefeeds in hostnames. We use \n as internal terminator. */
	if (memchr(&cmd[cmdlen + 1], '\n', len - cmdlen) != NULL)
		return got_error(GOT_ERR_BAD_PACKET);

	*hostname = strdup(&cmd[cmdlen + 1]);
	if (*hostname == NULL)
		return got_error_from_errno("strdup");

	/* Deny an empty hostname. */
	if ((*hostname)[0] == '\0') {
		free(*hostname);
		*hostname = NULL;
		return got_error(GOT_ERR_BAD_PACKET);
	}

	/* Deny overlong hostnames ,*/
	if (len - cmdlen > _POSIX_HOST_NAME_MAX)
		return got_error_fmt(GOT_ERR_NO_SPACE,
		    "hostname length exceeds %d bytes", _POSIX_HOST_NAME_MAX);

	/*
	 * TODO: More hostname verification? In any case, the provided
	 * value will have to match a string obtained from gotwebd.conf.
	 */

	return NULL;
}

int
main(int argc, char *argv[])
{
	const struct got_error *error;
	const char *unix_socket_path;
	int sock = -1;
	struct sockaddr_un	 sun;
	char *gitcmd = NULL, *command = NULL, *repo_path = NULL;
	char *hostname = NULL;
	int do_weblogin = 0;

#ifndef PROFILE
	if (pledge("stdio recvfd unix unveil", NULL) == -1)
		err(1, "pledge");
#endif
	if (strcmp(argv[0], GOTWEBD_LOGIN_CMD) == 0) {
		if (argc != 1 && argc != 2)
			usage();
		unix_socket_path = getenv("GOTWEBD_LOGIN_SOCKET");
		if (unix_socket_path == NULL)
			unix_socket_path = GOTWEBD_LOGIN_SOCKET;
		error = apply_unveil(unix_socket_path);
		if (error)
			goto done;
		if (argc == 2) {
			hostname = strdup(argv[1]);
			if (hostname == NULL) {
				error = got_error_from_errno("strdup");
				goto done;
			}
		}
		do_weblogin = 1;
	} else if (strcmp(argv[0], GOT_DIAL_CMD_SEND) == 0 ||
	    strcmp(argv[0], GOT_DIAL_CMD_FETCH) == 0) {
		if (argc != 2)
			usage();
		unix_socket_path = getenv("GOTD_UNIX_SOCKET");
		if (unix_socket_path == NULL)
			unix_socket_path = GOTD_UNIX_SOCKET;
		error = apply_unveil(unix_socket_path);
		if (error)
			goto done;
		if (asprintf(&gitcmd, "%s %s", argv[0], argv[1]) == -1)
			err(1, "asprintf");
		error = got_dial_parse_command(&command, &repo_path, gitcmd);
		if (error) {
			if (error->code == GOT_ERR_BAD_PACKET)
				usage();
			goto done;
		}
	} else if (argc == 3 && strcmp(argv[1], "-c") == 0) {
		if (strncmp(argv[2], GOTWEBD_LOGIN_CMD,
		    strlen(GOTWEBD_LOGIN_CMD)) == 0) {
			unix_socket_path = getenv("GOTWEBD_LOGIN_SOCKET");
			if (unix_socket_path == NULL)
				unix_socket_path = GOTWEBD_LOGIN_SOCKET;
			error = apply_unveil(unix_socket_path);
			if (error)
				goto done;
			error = parse_weblogin_command(&hostname, argv[2]);
			if (error) {
				if (error->code == GOT_ERR_BAD_PACKET)
					usage();
				goto done;
			}
			do_weblogin = 1;
		} else {
			unix_socket_path = getenv("GOTD_UNIX_SOCKET");
			if (unix_socket_path == NULL)
				unix_socket_path = GOTD_UNIX_SOCKET;
			error = apply_unveil(unix_socket_path);
			if (error)
				goto done;
			error = got_dial_parse_command(&command, &repo_path,
			    argv[2]);
			if (error) {
				if (error->code == GOT_ERR_BAD_PACKET)
					usage();
				goto done;
			}
		}
	} else
		usage();

	if ((sock = socket(AF_UNIX, SOCK_STREAM, 0)) == -1)
		err(1, "socket");

	memset(&sun, 0, sizeof(sun));
	sun.sun_family = AF_UNIX;
	if (strlcpy(sun.sun_path, unix_socket_path,
	    sizeof(sun.sun_path)) >= sizeof(sun.sun_path))
		errx(1, "gotd socket path too long");
	if (connect(sock, (struct sockaddr *)&sun, sizeof(sun)) == -1)
		err(1, "connect: %s", unix_socket_path);

	if (do_weblogin) {
#ifndef PROFILE
		if (pledge("stdio", NULL) == -1)
			err(1, "pledge");
#endif
		error = weblogin(stdout, sock, hostname);
	} else {
#ifndef PROFILE
		if (pledge("stdio recvfd", NULL) == -1)
			err(1, "pledge");
#endif
		error = got_serve(STDIN_FILENO, STDOUT_FILENO, command,
		    repo_path, sock, chattygot);
	}
done:
	free(gitcmd);
	free(command);
	free(repo_path);
	free(hostname);
	if (sock != -1)
		close(sock);
	if (error) {
		fprintf(stderr, "%s: %s\n", getprogname(), error->msg);
		return 1;
	}

	return 0;
}
